net/utils/net_snoop.c

```diff
@@ -161,76 +161,128 @@ begin_packed_struct struct snoop_packet_header_s
-        header.flags = htobe32(flags);
+        header->ts_usec = htobe64(SNOOP_EPOCH_USEC(tv));
+        header->flags = htobe32(flags);
         break;
-      }
-
-    case SNOOP_DATALINK_TYPE_TOKENBUS:
-    case SNOOP_DATALINK_TYPE_TOKERING:
-    case SNOOP_DATALINK_TYPE_METRONET:
-    case SNOOP_DATALINK_TYPE_ETHERNET:
-    case SNOOP_DATALINK_TYPE_HDLC:
-    case SNOOP_DATALINK_TYPE_CHARSYNC:
-    case SNOOP_DATALINK_TYPE_IBMC2C:
-    case SNOOP_DATALINK_TYPE_FDDI:
-    case SNOOP_DATALINK_TYPE_OTHER:
-      {
-        struct timeval tv;
 
+      case SNOOP_DATALINK_TYPE_TOKENBUS:
+      case SNOOP_DATALINK_TYPE_TOKERING:
+      case SNOOP_DATALINK_TYPE_METRONET:
+      case SNOOP_DATALINK_TYPE_ETHERNET:
+      case SNOOP_DATALINK_TYPE_HDLC:
+      case SNOOP_DATALINK_TYPE_CHARSYNC:
+      case SNOOP_DATALINK_TYPE_IBMC2C:
+      case SNOOP_DATALINK_TYPE_FDDI:
+      case SNOOP_DATALINK_TYPE_OTHER:
         gettimeofday(&tv, NULL);
-        header.ts.ts_sec = htobe32(tv.tv_sec);
-        header.ts.ts_usec = htobe32(tv.tv_usec);
-        header.rec_len = htobe32(flags);
+        header->ts.ts_sec = htobe32(tv.tv_sec);
+        header->ts.ts_usec = htobe32(tv.tv_usec);
+        header->rec_len = htobe32(flags);
         break;
-      }
 
-    default:
-      {
-        return -EINVAL;
-      }
+      default:
+        DEBUGASSERT(false);
     }
 
-  header.orig_len = htobe32(bytes);
-  header.incl_len = htobe32(bytes);
-  header.cum_drops = htobe32(drops);
+  header->orig_len = htobe32(bytes);
+  header->incl_len = htobe32(bytes);
+  header->cum_drops = htobe32(drops);
+}
 
-  ret = file_write(&snoop->filep, &header, sizeof(header));
-  if (ret != sizeof(header))
+/****************************************************************************
+ * Name: snoop_flush
+ *
+ * Description:
+ *   This function could flush snoop buf into file.
+ *
+ ****************************************************************************/
+
+static int snoop_flush(FAR struct snoop_s *snoop)
+{
+  ssize_t ret;
+
+  if (snoop->next == 0)
     {
-      return ret < 0 ? ret : -EINVAL;
+      return 0;
     }
 
-  return OK;
+  do
+    {
+      ret = file_write(&snoop->filep, snoop->buf, snoop->next);
+      if (ret < 0)
+        {
+          break;
+        }
+
+      snoop->next -= ret;
+      memmove(snoop->buf, snoop->buf + ret, snoop->next);
+    }
+  while (snoop->next > 0);
+
+#ifndef CONFIG_DISABLE_MOUNTPOINT
+  if (snoop->autosync)
+    {
+      ret = file_fsync(&snoop->filep);
+    }
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: snoop_flush_lock
+ *
+ * Description:
+ *   Snoop flush atomic
+ *
+ ****************************************************************************/
+
+static int snoop_flush_lock(FAR struct snoop_s *snoop)
+{
+  irqstate_t flags;
+  int ret;
+
+  flags = enter_critical_section();
+  nxmutex_lock(&snoop->mutex);
```

Comment:
why use both critical section and nxmutex

Reply:
snoop_flush could be called by interrupt context or app context. 
when in interrupt context case, snoop_flush involves with file sync and write operations，which could switch from interrupt context to app context, causing data race. 
so here need enter_critical_section and nxmutex_lock double check